pennylane/devices/device_api.py

```diff
@@ -0,0 +1,447 @@
+
+        1) Quantum Script in batch
+        2) Shot choice in a shot vector
+        3) Measurement in the quantum script
+        4) Parameter broadcasting
+        5) Measurement shape for array-valued measurements like probabilities
+
+        For a batch of quantum scripts with multiple measurements, a shot vector, and parameter broadcasting:
+
+        * ``result[0]``: the results for the first script
+        * ``result[0][0]``: the first shot number in the shot vector
+        * ``result[0][0][0]``: the first measurement in the quantum script
+        * ``result[0][0][0][0]``: the first parameter broadcasting choice
+        * ``result[0][0][0][0][0]``: the first value for an array-valued measurement
+
+        With the exception of quantum script batches, dimensions with only a single component should be eliminated.
+
+        For example:
+
+        With a single script and a single measurement process, execute should return just the
+        measurement value in a numpy array. ``shape`` currently accepts a device, as historically devices
+        stored shot information. In the future, this method will accept an ``ExecutionConfig`` instead.
+
+        >>> qs = qml.tape.QuantumScript([], [qml.expval(qml.PauliZ(0))])
+        >>> qs.shape(dev)
+        ()
+        >>> dev.execute(qs)
+        array(1.0)
+
+        If execute recieves a batch of scripts, then it should return a tuple of results:
+
+        >>> dev.execute([qs, qs])
+        (array(1.0), array(1.0))
+        >>> dev.execute([qs])
+        (array(1.0),)
+
+        If the script has multiple measurments, then the device should return a tuple of measurements.
+
+        >>> qs = qml.tape.QuantumScript([], [qml.expval(qml.PauliZ(0)), qml.probs(wires=(0,1))])
+        >>> qs.shape(dev)
+        ((), (4,))
+        >>> dev.execute(qs)
+        (array(1.0), array([1., 0., 0., 0.]))
+
+        """
+        raise NotImplementedError
+
+    @classmethod
+    def supports_jacobian_with_configuration(cls, execution_config=None) -> bool:
+        """Determine whether or not a gradient is available with a given execution configuration.
+
+        Default behaviour assumes first order derivatives exist if :meth:`~.gradient` is overriden.
+
+        Args:
+            execution_config (ExecutionConfig): A description of the hyperparameters for the desired computation.
+
+        Returns:
+            Bool
+
+        **Example:**
+
+        For example, the current Python device supports adjoint differentiation if the order is ``1`` and
+        the execution occurs with no shots ``shots=None``.
+
+        >>> config = ExecutionConfig(order=1, shots=None)
+        >>> dev.supports_gradient_with_configuration(config)
+        True
+        >>> config = ExecutionConfig(order=1, shots=10)
+        >>> dev.supports_gradient_with_configuration(config)
+        False
+        >>> config = ExecutionConfig(order=2, shots=None)
```

Comment:
If the execution config doesn't specify _how_ the gradient is requested to be computed, what does this method assume?

Reply:
```suggestion
        >>> config = ExecutionConfig(order=2, shots=None, gradient_method="device")
```